src/NServiceBus.Core/Transports/Msmq/ReceiveOnlyWithNativeTransaction.cs

```diff
@@ -0,0 +1,110 @@
+﻿namespace NServiceBus
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Messaging;
+    using System.Threading.Tasks;
+    using Transport;
+
+    class ReceiveOnlyWithNativeTransaction : ReceiveStrategy
+    {
+        public ReceiveOnlyWithNativeTransaction(MsmqFailureInfoStorage failureInfoStorage)
+        {
+            this.failureInfoStorage = failureInfoStorage;
+        }
+
+        public override async Task ReceiveMessage()
+        {
+            Message message = null;
+
+            try
+            {
+                using (var msmqTransaction = new MessageQueueTransaction())
+                {
+                    msmqTransaction.Begin();
+
+                    if (!TryReceive(msmqTransaction, out message))
+                    {
+                        return;
+                    }
+
+                    Dictionary<string, string> headers;
+
+                    if (!TryExtractHeaders(message, out headers))
+                    {
+                        MovePoisonMessageToErrorQueue(message, IsQueuesTransactional ? MessageQueueTransactionType.Single : MessageQueueTransactionType.None);
+
+                        msmqTransaction.Commit();
+                        return;
+                    }
+
+                    var shouldCommit = await ProcessMessage(message, headers).ConfigureAwait(false);
+
+                    if (shouldCommit)
+                    {
+                        msmqTransaction.Commit();
+                    }
+                    else
+                    {
+                        msmqTransaction.Abort();
+                    }
+                }
+            }
+            // We'll only get here if Commit/Abort/Dispose throws which should be rare.
+            // Note: If that happens the attempts counter will be inconsistent since the message might be picked up again before we can register the failure in the LRU cache.
+            catch (Exception exception)
+            {
+                if (message == null)
+                {
+                    throw;
+                }
+
+                failureInfoStorage.RecordFailureInfoForMessage(message.Id, exception);
+            }
+        }
+
+        async Task<bool> ProcessMessage(Message message, Dictionary<string, string> headers)
+        {
+            var transportTransaction = new TransportTransaction();
```

Comment:
Since the `TransportTransaction` isn't used for anything here, you could allocate a single one and reuse it instead of allocating one per message. 

Same thing would be true for the `ContextBag` in the `ReceiveStrategy.TryProcessMessage` call, though probably out of scope for this PR.